### PR TITLE
Return complete list of DDL commands for creating foreign keys in SQLite

### LINKED ISSUE #3351: Allow creating foreign keys in sqlite
# Environment

Knex version: 0.18.3
Database + version: sqlite3 v4.0.7
OS: macos

# Bug / Feature request

Currently knex complaints about adding a foreign key when altering a table (`SQLite3 Foreign & Primary keys may only be added on create`). I understand why you have it there, but it is not 100% true. You can create foreign keys on existing table - you just have to add new column with that key, the thing you cannot do is to alter existing column (making it foreign key). 

Currently I am trying to use knex as a query builder/runner in my ORM (https://github.com/B4nan/mikro-orm), and the way I am talking about is how it works there currently (so it is not theoretical, it is something I already have implemented without knex - see here https://github.com/B4nan/mikro-orm/tree/master/lib/schema). 

Here is example SQL that is valid and I would like to be able to build that with knex schema builder:

```sql
ALTER TABLE "author" ADD "favourite_book_id" INTEGER DEFAULT NULL REFERENCES "book" ("id") ON DELETE SET NULL ON UPDATE CASCADE;
```

The reason why I don't want to create this FK when creating the table is that there are references in both ways - author table references book and vice versa. It is a model example, not a real world usage.

### PR DESCRIPTION
**Pull request by kibertoad**
fixes #3351

### CODE REVIEW

**Comment by kibertoad** [comment]
@B4nan Could you please review? If this works for you, I could revise all other SQLite commands to follow the same pattern.

File: lib/dialects/sqlite3/schema/internal/sqlite-ddl-operations.js [hunk 1]
```diff
@@ -1,21 +1,34 @@
...
+    );
   }
+  return result;
 }
 
 function createTempTable(createTable, tablename, alteredName) {
   return createTable.sql.replace(tablename, alteredName);
 }
 
 async function copyData(trx, tableName, alteredName) {
```
Comment (kibertoad) [inline lib/dialects/sqlite3/schema/internal/sqlite-ddl-operations.js:20]: `copyData` and `reinsertData` will be removed in the end, we'll use `copyAllData` for everything, as it's significantly better for large amount of data.

File: test/integration2/schema/foreign-keys.spec.js [hunk 2]
```diff
@@ -37,14 +40,71 @@ describe('Schema', () => {
...
+              'foreign_keys_table_one',
+              (table) => {
+                table
+                  .foreign('fkey_three')
+                  .references('foreign_keys_table_three.id')
+                  .withKeyName('fk_fkey_threeee');
+              }
+            );
+            const queries = await builder.generateDdlCommands();
+            expect(queries).to.eql([
```
Comment (kibertoad) [inline test/integration2/schema/foreign-keys.spec.js:65]: @B4nan This is the format that you would be getting. Would that work?

**Comment by nickrum** [comment]
Looks very promising. It is unfortunate though that it is not possible to return the generated commands directly from `toSQL()`. @kibertoad Have you considered changing that function to return a Promise? (Although that would be a rather heavy breaking change)

**Comment by kibertoad** [comment]
Yes, exactly. Heavily breaking change and one not needed for the majority of cases.

**Comment by kibertoad** [comment]
Hence the new metod generateDdlCommands.

**Comment by B4nan** [comment]
Finally had some time to review this, it looks good but it there are some issues:

- sometimes the meta queries are there (like `pragma ...`), which I am not interested in
- other drivers need to be probably adjusted too, like in mysql the `renameColumn` now prints `show fields from `author` where field = ?;` instead of the actual rename query. also looking at the implementation I would very much like just a rename column query, while it tries to be too smart and fix FKs and indexes too (which I have already handled myself just fine, and wont to have control over those) - could it be at least configurable?
- the TS typings are wrong, it says there will be array of strings but its actually array of objects with `sql` and `bindings` keys
- what is the exact purpose of the `pre` and `post` arrays? so far i never seen anything filled in there. would be great if those could be used for the "too smart" queries i mentioned about the mysql's rename column method, so i can just ignore them this way

**Pull request merged**